dma: intel_adsp: Intel cAVS derivative of dma_dw

teburd opened the pull request:
Expose most of the existing DesignWare DMA driver as sharable, common code,
for derivations of the IP such as the one used in intel_adsp
as a general purpose dma controller. An added init step to disable dynamic clocking
is needed prior to manipulating the IP configuration registers for intel_adsp.

The data and config structs of cavs_gpdma are laid out to effectively
be extensions to the dma_dw config and data structs so pointer aliasing
works.

Next step improvements would be enabling and using linear link position
tracking and enabling device power management which are also intel_adsp specifics.

A few improvements to the common code are needed to support LLI (linked list block transfer descriptions) transfers where
the hardware will run through a list of transfer descriptions by itself.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

andyross commented on drivers/dma/dma_cavs_gpdma.c:
```diff
@@ -0,0 +1,137 @@
+
+#define CAVS_GPDMA_INIT(inst)						\
+									\
+	static struct dw_drv_plat_data dmac##inst = {			\
```
Seems like you could get the number of channels out of devicetree and then just declare this as an array.  I know various instantiations are intended to have different numbers of channels.

Also as the initialization values are identical in all cases, this seems like a waste of bytes in .data vs. just putting it in .bss and initializing at runtime by setting class to 6.

teburd replied:
They aren't always the same in SOF, and this is something I need to port over. They vary slightly per platform

andyross commented on drivers/dma/dma_cavs_gpdma.c:
```diff
@@ -0,0 +1,137 @@
+};
+
+/* Disables automatic clock gating (force disable clock gate) */
+static void cavs_gpdma_clock_enable(const struct device *dev)
```
Since it looks like the only delta vs. the main driver is this clock enable step, another design option here would be to use the existing driver unaltered and put the clock handling into a power management callback set up at the soc level.

teburd replied:
There's other alterations needed once I add linear link positioning and linked list transfers as well as some individual hardware soc workarounds I need to port over from sof. Some of that will be here, some of that will be in the common bits with some ifdefry (has to be I think)

andyross commented on drivers/dma/dma_cavs_gpdma.c:
```diff
@@ -0,0 +1,137 @@
+	.stop = dw_dma_transfer_stop,
+};
+
+#define CAVS_GPDMA_INIT(inst)						\
```
This is a BIG macro.  Maybe it might be cleaner to do several FOREACH iterations over the devicetree instances, e.g. one to define the config structs, one to declare interrupts, etc...?

teburd replied:
That's a fair statement, mostly people write the one macro to init all the things for each instance, but this is quite big. Perhaps splitting it up at least might make it easier to read through, and perhaps share more with the original dw_dma.c

andyross commented on drivers/dma/dma_cavs_gpdma.c:
```diff
@@ -0,0 +1,137 @@
+			      CONFIG_DMA_INIT_PRIORITY,			\
+			      &cavs_gpdma_driver_api);			\
+									\
+	static void cavs_gpdma##inst##_irq_config(void)			\
```
Why not just put the irq into the config struct as a field and use a single function?  The IRQ_CONNECTS() for all instances can then be put into a much smaller FOREACH in a global sysinit or something.  (I mean, I agree that this is a spot where IRQ_CONNECT()'s inability to be declared statically sort of bites us, but I'm not sure defining a separate nearly-noop function for every instance is the cleanest workaround.)

teburd replied:
IRQ_CONNECT boils down to, as I recall, a table in static memory in a specified linker section for xtensa. I can't use a struct variable, even if const and static itself (its not a const expression for the macro expansion) to dynamically do IRQ_CONNECT(). I could use irq_connect_dynamic with a common function and save only the irq num here. It's not necessarily the common thing to do as on other platforms IRQ_CONNECT puts things in flash I believe to save memory on the vector table. Does that matter here? probably not right.

teburd replied:
I misunderstood the point here last week. I get it now, but to be fair this is the way the dma dw driver was written previously. I could do the above but I actually prefer to see the CONNECT and enable colocated here as it makes it easy to understand. I could do what I initially thought you meant, and save the irq num and swap to irq_connect_dynamic here with a single function but I'm fine with the way it is.

It would be nice if IRQ_CONNECT could be statically declared, and looking at the macros again it seems like it really could be statically declared for Xtensa at least. I pushed a commit that does just that and uses it here. This commit could be dropped if its not liked.

andyross reviewed (dismissed):
Some nitpicking, but nothing that should block merge.  Is there a test for this we can use on this hardware?

teburd commented:
> Some nitpicking, but nothing that should block merge. Is there a test for this we can use on this hardware?

Yes, the entire tests/drivers/dma test suite passes (only verifies gpdma0 though, I need to fix that)

kv2019i commented on drivers/dma/dma_dw_common.c:
```diff
@@ -0,0 +1,333 @@
+
+	uint32_t status_tfr = 0U;
+	uint32_t status_block = 0U;
+	uint32_t status_err = 0U;
```
Not added in this patch, but: unnecessary init of "status_err", it's initialized on L62.

kv2019i commented on soc/xtensa/intel_adsp/cavs_v25/Kconfig.defconfig.series:
```diff
@@ -72,4 +72,8 @@ config KERNEL_VM_SIZE
+config DMA_CAVS_GPDMA
+	default y
+	depends on DMA
+
```
This is going to be interesting how we manage the transition. In the current SOF build, this defconfig file is used, but currently Zephyr CONFIG_DMA is not set, so this gets ignored (and the DWDMA driver from SOF is used instead). But something to watch out in the future.

kv2019i reviewed (commented):
In general looks good. I'll let you the Zephyr experts conclude on the DT definitions. A few minor notes.

dcpleung commented:
Recommend dividing these into smaller commits.

Commit:
dma: dw: Common dw dma driver functionality

Move most of the designware driver into a common compile unit with a
a header that exposes the common functionality.

This allows for derivative hardware, such as that in intel's adsp (cavs)
to use the common functionality while extending.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
dma: cavs: Add gpdma derivative of dw dma for cavs

Intel's adsp needs to set, at a minimum, a clocking bit before the driver
can initialize the designware dma controller. In many ways it is the
designware dmac IP but with additional registers and functionality added
on top of it. So the code structure here follows how the hardware
appears to be designed, layered on top of the designware driver.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
dma: intel_adsp: Reduce irq setup code

Makes the IRQ_CONNECT macro statically declarable for Xtensa (nothing
prevented it except its structuring). Allows for IRQ_CONNECT usage
on a xtensa only platform to be declared statically avoiding the
fluff of setting up and enabling irqs for cavs gpdma.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>
